File: Day21/PROG_타일장식물_200407.java
/*
 * programmers lv3 타일 장식물 
 */

public class PROG_타일장식물_200407 {
	public static void main(String[] args) {
		System.out.println(solution(5));
		System.out.println(solution(6));
	}

	public static long solution(int N) {
		long answer = 0;
		long[] f = new long[N+1];
		f[1] = 1;
		f[2] = 1;
		for (int i = 3; i <= N; i++) {
			f[i] = f[i-2] + f[i-1];
		}
		
		answer = f[N]*2 + (f[N]+f[N-1])*2;
		
		return answer;
	}
}

File: PROG_타일장식물_200407.cpp
/* 타일 장식물 Level 3
https://programmers.co.kr/learn/courses/30/lessons/43104
 */
#include <string>
#include <vector>
#include <iostream>
using namespace std;

long long solution(int N) {
    if (N == 1) return 4;
    vector<long long> fibo(N+1);
    fibo[0] = fibo[1] = 1;
    for (int i=2; i<=N; i++) {
        fibo[i] = fibo[i-1]+fibo[i-2];
    }
    return fibo[N]*2 + fibo[N-1]*2;
}

int main() {
    cout << solution(5) << '\n' << solution(6);
}
